azurelinuxagent/common/protocol/wire.py

```diff
@@ -780,70 +780,88 @@ def update_host_plugin_from_goal_state(self):
 
-                    if goal_state_updated or vm_settings_goal_state_updated:
-                        # compare() raises a GoalStateMismatchError if the goal states don't match
-                        ExtensionsGoalState.compare(self._goal_state.extensions_config, vm_settings_goal_state)
-
-                    vm_settings_goal_state_is_valid = True
-
                 except Exception as error:
                     # _fetch_vm_settings_goal_state() does its own detailed error reporting and raises ProtocolError; do not report those
                     if not isinstance(error, ProtocolError):
                         self._vm_settings_error_reporter.report_error(format_exception(error))
                 self._vm_settings_error_reporter.report_summary()
 
             #
-            # If we fetched a valid vmSettings, use that for the extensions goal state. Otherwise use ExtensionsConfig
+            # 3) Lastly we fetch the rest of the goal state from the WireServer (but ony if needed: initialization a "forced" update or
+            #    a change in the incarnation). Note that if we fetch the full goal state we also update self._goal_state.
+            #
+            if force_update:
+                logger.info("Forcing an update of the goal state..")
+
+            fetch_full_goal_state = force_update or self._goal_state is None or self._goal_state.incarnation != goal_state.incarnation
+
+            if not fetch_full_goal_state:
+                goal_state_updated = False
+            else:
+                goal_state.fetch_full_goal_state(self)
+                self._goal_state = goal_state
+                goal_state_updated = True
+
+            #
+            # If we fetched the vmSettings compare them against extensionsConfig and use them for the extensions goal state if
+            # everything matches, otherwise use extensionsConfig.
             #
-            if vm_settings_goal_state_is_valid:
+            use_vm_settings = False
+            if vm_settings_goal_state is not None:
+                if not goal_state_updated and not vm_settings_goal_state_updated:  # no need to compare them, just use vmSettings
+                    use_vm_settings = True
+                else:
+                    try:
+                        ExtensionsGoalState.compare(self._goal_state.extensions_config, vm_settings_goal_state)
+                        use_vm_settings = True
+                    except GoalStateMismatchError as mismatch:
+                        if not is_retry and mismatch.attribute in ("created_on_timestamp", "activity_id"):
+                            # this may be OK; a new goal state may have arrived in-between the calls to the HostGAPlugin and the WireServer;
+                            # retry one time after and then report the error if it happens again
+                            time.sleep(conf.get_goal_state_period())
```

Comment:
actually i was thinking of harcoding it 2 seconds, but at the end went for the goal state period. I do not think that initial goal state is relevant here, since this is a retry